src/ppx/grfx/vk/vk_shading_rate.cpp

```diff
@@ -0,0 +1,315 @@
+    imageCreateInfo.height = (pCreateInfo->framebufferSize.height + mTexelSize.height - 1) / mTexelSize.height;
+
+    PPX_CHECKED_CALL(GetDevice()->CreateImage(&imageCreateInfo, &mAttachmentImage));
+
+    VkImageViewCreateInfo vkci           = {VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO};
+    vkci.flags                           = 0;
+    vkci.image                           = ToApi(mAttachmentImage)->GetVkImage();
+    vkci.viewType                        = VK_IMAGE_VIEW_TYPE_2D;
+    vkci.format                          = ToVkFormat(imageCreateInfo.format);
+    vkci.subresourceRange.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+    vkci.subresourceRange.baseMipLevel   = 0;
+    vkci.subresourceRange.levelCount     = 1;
+    vkci.subresourceRange.baseArrayLayer = 0;
+    vkci.subresourceRange.layerCount     = 1;
+
+    VkResult vkres = vk::CreateImageView(
+        ToApi(GetDevice())->GetVkDevice(),
+        &vkci,
+        nullptr,
+        &mAttachmentView);
+    if (vkres != VK_SUCCESS) {
+        PPX_ASSERT_MSG(false, "vkCreateImageView(ShadingRatePatternView) failed: " << ToString(vkres));
+        return ppx::ERROR_API_FAILURE;
+    }
+
+    return ppx::SUCCESS;
+}
+
+void ShadingRatePattern::DestroyApiObjects()
+{
+    if (mAttachmentView) {
+        vkDestroyImageView(
+            ToApi(GetDevice())->GetVkDevice(),
+            mAttachmentView,
+            nullptr);
+        mAttachmentView.Reset();
+    }
+    if (mAttachmentImage) {
+        GetDevice()->DestroyImage(mAttachmentImage);
+        mAttachmentImage.Reset();
+    }
+}
+
+std::unique_ptr<ShadingRatePattern::RenderPassModifier> ShadingRatePattern::CreateRenderPassModifier(vk::Device* device, ShadingRateMode mode)
+{
+    switch (mode) {
+        case SHADING_RATE_FDM:
+            return std::make_unique<ShadingRatePattern::FDMRenderPassModifier>();
+        case SHADING_RATE_VRS:
+            return std::make_unique<ShadingRatePattern::VRSRenderPassModifier>(device->GetShadingRateCapabilities());
+        default:
+            return nullptr;
+    }
+}
+
+VkAttachmentReference2 ToVkAttachmentReference2(const VkAttachmentReference& ref)
+{
+    VkAttachmentReference2 ref2 = {VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_2};
+    ref2.attachment             = ref.attachment;
+    ref2.layout                 = ref.layout;
+    return ref2;
+}
+
+std::vector<VkAttachmentReference2> ToVkAttachmentReference2(uint32_t referenceCount, const VkAttachmentReference* pReferences)
+{
+    std::vector<VkAttachmentReference2> references2(referenceCount);
+    for (uint32_t i = 0; i < referenceCount; ++i) {
+        references2[i] = ToVkAttachmentReference2(pReferences[i]);
+    }
+    return references2;
+}
+
+void ShadingRatePattern::RenderPassModifier::Initialize(const VkRenderPassCreateInfo& vkci)
+{
+    auto& vkci2 = mVkRenderPassCreateInfo2;
+    vkci2.pNext = vkci.pNext;
+    vkci2.flags = vkci.flags;
+
+    mAttachments.clear();
+    mAttachments.reserve(vkci.attachmentCount + 1);
```

Comment:
The `+1` is for the `VkAttachmentDescription2` added for the denisty map or vrs attachment.

Reply:
Right, in the other function, the `push_back`.
Right.
What I'm not a fan of is we have attributes which are initialized across multiple functions, but function are not isolated at all. If you remove this reserve in one function, it brings a nasty bug in another function when we push-back and potentially end-up with bad pointers.

In understand why you split it, as each override will have to initialize part of the structure differently, but there might be something better to do.
Something where the vector is passed around, and the scope where each element is known to be fully initialized is a bit clearer.

Maybe something like:
```cpp
mAttachments = GetBaseAttachments();
auto additionnal_attachments_it = AddImplementationAttachments(attachments);

// Calls the base function to initialize create info, then internally calls the implem specific function. But it's a const method.
// The passed iterator tells us where the implementation specific attachments are.
mCreateInfo = MakeCreateInfo(attachments, additionnal_attachments_it);
```

Reply:
>  If you remove this reserve in one function, it brings a nasty bug in another function when we push-back and potentially end-up with bad pointers.

I believe the `reserve()` call here is a very minor optimization, not critical. The `push_back()` call in `InitializeImpl()` might reallocate the vector memory, but the `pAttachments` will be updated to point to the new memory.

Since this makes the code harder to understand and provides very little benefit, I've removed the `reserve()` call.

> Something where the vector is passed around, and the scope where each element is known to be fully initialized is a bit clearer.

I don't think I fully understand the suggestion. The fact that `InitializeImpl()` modifies `mAttachments` is an implementation detail that I don't think needs to be exposed. All of the pointers are valid both before and after the `InitializeImpl()` call.

I did some refactoring/renaming/commenting that I hope makes this code more clear. If you still think the change you were suggesting makes sense, please clarify.

Reply:
Thanks for the refactoring!